File: examples/javascript/lib/address.js
// Addresses in an invitation's `at`, as CARRIER-TCP.md and CARRIER-WEB.md write them.

// A host as RFC 3986 writes one: a registered name, an IPv4 address, or an IPv6 address in brackets.
const HOST = String.raw`(\[[0-9A-Fa-f:.]+\]|[A-Za-z0-9\-._~!$&'()*+,;=%]+)`;
// The scheme is read without regard to case, and nothing follows the port.
const TCP = new RegExp(String.raw`^tcp://${HOST}:(\d+)$`, "i");

// The schemes this kit dials, and those it listens on.
export const DIALED = ["tcp", "http", "https", "ws", "wss"];
export const LISTENED = ["tcp", "http", "ws"];

// Reads one address. Returns { scheme, host, port, url } or null when it is no address this kit dials.
export function readAddress(s) {
  if (typeof s !== "string") return null;
  const m = /^([A-Za-z][A-Za-z0-9+.-]*):/.exec(s);
  if (!m) return null;
  const scheme = m[1].toLowerCase();
  if (scheme === "tcp") {
    const t = TCP.exec(s);
    if (!t) return null;
    const port = Number(t[2]);
    if (port < 1 || port > 65535) return null;
    const host = t[1].startsWith("[") ? t[1].slice(1, -1) : t[1];
    if (t[1].startsWith("[") && !URL.canParse(`http://${t[1]}/`)) return null;
    return { scheme, host, port, url: s };
  }
  if (!DIALED.includes(scheme)) return null;
  // RFC 3986: only its characters, and an authority that is not empty.
  if (!/^[A-Za-z0-9\-._~:/?#[\]@!$&'()*+,;=%]+$/.test(s) || !/^[^:]+:\/\/[^/?#]/.test(s)) return null;
  let u;
  try {
    u = new URL(s);
  } catch {
    return null;
  }
  // With a host, and with no user information and no fragment.
  if (!u.hostname || s.includes("#") || u.username || u.password || /^[^/]*\/\/[^/?]*@/.test(s)) return null;
  return { scheme, url: s };
}

// The addresses of `at` this kit dials, in the order written.
export const dialable = (at) => (Array.isArray(at) ? at.map(readAddress).filter(Boolean) : []);

File: examples/go/address.go
package main

import (
	"net"
	"net/url"
	"strconv"
	"strings"
)

// An address is a URI of RFC 3986 whose scheme names a carrier this kit
// stands, written as that carrier writes it.
type address struct {
	text     string // as written, what the post is sent to
	form     string // "tcp", "post" or "line"
	secure   bool   // https and wss carry the bytes inside TLS
	hostPort string // what a socket dials
	url      *url.URL
}

// forms is every scheme this kit stands, and the form each names.
var forms = map[string]string{"tcp": "tcp", "http": "post", "https": "post", "ws": "line", "wss": "line"}

// uriChar is every character RFC 3986 lets a URI hold outside a
// percent-encoding.
func uriChar(c byte) bool {
	return c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z' || c >= '0' && c <= '9' ||
		strings.IndexByte("-._~:/?#[]@!$&'()*+,;=", c) >= 0
}

func isHexDigit(c byte) bool {
	return c >= '0' && c <= '9' || c >= 'a' && c <= 'f' || c >= 'A' && c <= 'F'
}

// parseAddress reads one address. It answers false for a string that is not
// a URI with a scheme, for a scheme this kit does not stand, and for an
// address its carrier does not write.
func parseAddress(s string) (*address, bool) {
	for i := 0; i < len(s); i++ {
		switch {
		case s[i] == '%':
			if i+2 >= len(s) || !isHexDigit(s[i+1]) || !isHexDigit(s[i+2]) {
				return nil, false
			}
		case !uriChar(s[i]):
			return nil, false
		}
	}
	u, err := url.Parse(s)
	if err != nil || u.Opaque != "" || u.User != nil || strings.Contains(s, "#") || u.Hostname() == "" {
		return nil, false
	}
	form, ok := forms[u.Scheme]
	if !ok {
		return nil, false
	}
	a := &address{text: s, form: form, secure: u.Scheme == "https" || u.Scheme == "wss", url: u}
	port := u.Port()
	if form == "tcp" {
		// tcp://host:port, and nothing after the port
		if u.Path != "" || u.RawQuery != "" || u.ForceQuery || port == "" {
			return nil, false
		}
	} else if port == "" {
		port = "80"
		if a.secure {
			port = "443"
		}
	}
	if n, err := strconv.Atoi(port); err != nil || len(port) > 5 || n < 1 || n > 65535 {
		return nil, false
	}
	a.hostPort = net.JoinHostPort(u.Hostname(), port)
	return a, true
}
